src/System.Diagnostics.Process/src/System/Diagnostics/AsyncStreamReader.cs

```diff
@@ -148,76 +88,64 @@ internal void BeginReadLine()
-
-            try
-            {
-                byteLen = t.GetAwaiter().GetResult();
-            }
-            catch (IOException)
-            {
-                // We should ideally consume errors from operations getting cancelled
-                // so that we don't crash the unsuspecting parent with an unhandled exc. 
-                // This seems to come in 2 forms of exceptions (depending on platform and scenario), 
-                // namely OperationCanceledException and IOException (for errorcode that we don't 
-                // map explicitly).   
-                byteLen = 0; // Treat this as EOF
-            }
-            catch (OperationCanceledException)
-            {
-                // We should consume any OperationCanceledException from child read here  
-                // so that we don't crash the parent with an unhandled exc
-                byteLen = 0; // Treat this as EOF
-            }
-
-            if (byteLen == 0)
+            while (true)
             {
-                // We're at EOF, we won't call this function again from here on.
-                lock (_messageQueue)
+                try
                 {
-                    if (_sb.Length != 0)
-                    {
-                        _messageQueue.Enqueue(_sb.ToString());
-                        _sb.Length = 0;
-                    }
-                    _messageQueue.Enqueue(null);
-                }
+                    int bytesRead = await _stream.ReadAsync(_byteBuffer, 0, _byteBuffer.Length, _cts.Token).ConfigureAwait(false);
+                    if (bytesRead == 0)
+                        break;
 
-                try
+                    int charLen = _decoder.GetChars(_byteBuffer, 0, bytesRead, _charBuffer, 0);
+                    _sb.Append(_charBuffer, 0, charLen);
+                    MoveLinesFromStringBuilderToMessageQueue();
+                }
+                catch (IOException)
                 {
-                    // UserCallback could throw, we should still set the eofEvent 
-                    FlushMessageQueue();
+                    // We should ideally consume errors from operations getting cancelled
+                    // so that we don't crash the unsuspecting parent with an unhandled exc.
+                    // This seems to come in 2 forms of exceptions (depending on platform and scenario),
+                    // namely OperationCanceledException and IOException (for errorcode that we don't
+                    // map explicitly).
+                    break; // Treat this as EOF
                 }
-                finally
+                catch (OperationCanceledException)
                 {
-                    _eofEvent.Set();
+                    // We should consume any OperationCanceledException from child read here
+                    // so that we don't crash the parent with an unhandled exc
+                    break; // Treat this as EOF
                 }
+
+                FlushMessageQueue();
```

Comment:
There is now a subtle behavioral difference between the old and new implementations.  In the old implementation, any exceptions from within the callback would be eaten.  In the new implementation, any exceptions from within the callback will end up getting propagated out of WaitForEOF when the task is waited on. And both the old and the new behaviors are different from desktop, where any exception from the user callback will likely go unhandled on an arbitrary thread pool thread.  @pallavit, we should decide what the right behavior here is; if we want to match the previous .NET Core behavior, that's easy, we just wrap a try/catch around the body.  If we want to match the desktop behavior, we'd likely need to queue a work item that just re-throws the exception.